conan/external_tools/vswhere_test.py

```diff
@@ -0,0 +1,169 @@
+from conans.model import settings
+
+
+class vswhereTest(unittest.TestCase):
+
+    # Environment supossed:
+    #   - BuildTools 14 (2015)
+    #   - VS Community 14 (2015)
+    #
+    #   - BuildTools 15 (2017) OR VS Community 15 (2017)
+    modern_products = 1 # 2017 or higher versions -> vswhere()
+    all_modern_products = 2 # 2017 or higher versions -> vswhere(products=["*"])
+    modern_and_legacy_products = 2 # 2017 and lower versions -> vswhere(legacy=True)
+    only_legacy_products = 1
+    all_products = 3
+
+    def setUp(self):
+        if Version(client_version) < Version("1.1"):
+            raise nose.SkipTest('Only >= 1.1 version')
+
+    def vs_comntools_test(self):
+        # Fake path
+        print(tools.__file__)
+        with tools.environment_append({"VS150COMNTOOLS": "fake/path/here"}):
+            path = tools.vs_comntools("15")
+            self.assertEqual(path, "fake/path/here")
+
+        # VS 14 path
+        path = tools.vs_comntools("14")
+        self.assertEqual(path, "C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\Common7\\Tools\\")
+
+        # VS 15 path (shouldn't be found as VS150COMNTOOLS is not set by default)
+        path = tools.vs_comntools("15")
+        self.assertEqual(path, None)
+
+    def vswhere_test(self):
+        # test there is some output
+        output = tools.vswhere()
+        self.assertIsNotNone(output)
+
+        # products and legacy not allowed
+        self.assertRaises(ConanException, tools.vswhere, products=["*"], legacy=True)
+
+        # Detect only one product (VS Community 15) as vswhere default detection
+        nproducts = len(tools.vswhere())
+        self.assertEqual(nproducts, self.modern_products)
+
+        # Detect only modern products (VS Community 15 & BuildTools 15)
+        products = tools.vswhere(products=["*"])
+        nproducts = len(products)
+        
+        self.assertEqual(nproducts, self.all_modern_products)
+        for product in products:
+            install_path = product["installationPath"]
+            self.assertTrue("Community" in install_path or
+                            "BuildTools" in install_path)
+
+        # Detect also legacy products but no modern BuildTools
+        products = tools.vswhere(legacy=True)
+        nproducts = len(products)
+
+        self.assertEqual(nproducts, self.modern_and_legacy_products)
+        for product in products:
+            install_path = product["installationPath"]
+            self.assertTrue("Community" in install_path or
+                            "Microsoft Visual Studio 14.0" in install_path)
+
+        # Detect all installed products
+        products = tools.vswhere(products=["*"])
+        products += tools.vswhere(legacy=["*"])
+        seen_products = []
+        for product in products:
+            if product not in seen_products:
+                seen_products.append(product)
+        products = seen_products
+        nproducts = len(products)
```

Comment:
All this block from "seen_products" is very very confusing. I really don't know what are you trying to do. removing duplicates? you can do just `set(products)` and it will remove duplicated elements.

Reply:
not able to do a set(products) here because python says dict types are not hashable

Reply:
oh! they are dicts, well, then you can `set(products.keys())`

Reply:
is a list of dicts actually

Reply:
Do wanna you play uh? Could this work?

`products += [new_product for new_product in tools.vswhere(legacy=["*"]) if new_product not in products]`

Reply:
unfortunately it is not working... :(